src/stream/futures.rs

```diff
@@ -0,0 +1,83 @@
+use future::{Future, IntoFuture};
+use stream::Stream;
+use poll::Poll;
+use Async;
+use stack::{Stack, Drain};
+use std::sync::Arc;
+use task::{self, UnparkEvent};
+
+use std::prelude::v1::*;
+
+/// An adaptor for a stream of futures to execute the futures concurrently, if
+/// possible, delivering results as they become available.
+///
+/// This adaptor will return their results in the order that they complete.
+/// This is created by the `futures` method.
+///
+#[must_use = "streams do nothing unless polled"]
+pub struct Futures<F>
+    where F: Future
+{
+    futures: Vec<F>,
+    stack: Arc<Stack<usize>>,
+    pending: Option<Drain<usize>>,
+    active: usize,
+}
+
+/// Converts a `Vec` of `IntoFuture` elements into a `Stream` over the underlying
+/// futures. The created stream will yield into values produced by those futures.
+/// Similar to `buffer_unordered` the order of the produced items will not be equal
+/// with the order of elements in the input vector.
+///
+pub fn futures<I>(futures: I) -> Futures<<I::Item as IntoFuture>::Future>
+    where I: IntoIterator,
+          I::Item: IntoFuture
+{
+    let futures = futures.into_iter().map(IntoFuture::into_future).collect::<Vec<_>>();
+    let stack = Arc::new(Stack::new());
+    for i in 0..futures.len() {
+        stack.push(i);
+    }
+    Futures {
+        active: futures.len(),
+        futures: futures,
+        pending: None,
+        stack: stack,
+    }
+}
+
+impl<F> Stream for Futures<F>
+    where F: Future
+{
+    type Item = F::Item;
+    type Error = F::Error;
+
+    fn poll(&mut self) -> Poll<Option<Self::Item>, Self::Error> {
+        if self.active == 0 {
+            return Ok(Async::Ready(None));
+        }
+        let mut drain = if self.pending.is_some() {
+            ::std::mem::replace(&mut self.pending, None).unwrap()
+        } else {
+            self.stack.drain()
+        };
+        while let Some(id) = drain.next() {
+            let event = UnparkEvent::new(self.stack.clone(), id);
+            let ref mut f = self.futures[id];
+            match task::with_unpark_event(event, || f.poll()) {
+                Ok(Async::NotReady) => {}
```

Comment:
Perhaps this branch could be `continue` and the `active`/`pending` adjustments of the next two branches could be shared?